File: java/ToolGood/Words/internals/TrieNodeEx.java
package ToolGood.Words.internals;

import java.util.ArrayList;
import java.util.Hashtable;
import java.util.List;
import java.util.Map;

public class TrieNodeEx{
    public TrieNodeEx Parent;
    public TrieNodeEx Failure;
    public Character Char;
    public boolean End;
    public List<Integer> Results;
    public Map<Character, TrieNodeEx> m_values;
    public Map<Character, TrieNodeEx> merge_values;
    private Integer minflag = Integer.MAX_VALUE;
    private Integer maxflag = 0;
    public Integer Next;
    private Integer Count;


    public TrieNodeEx()
    {
        m_values = new Hashtable<Character, TrieNodeEx>();
        merge_values = new Hashtable<Character, TrieNodeEx>();
        Results = new ArrayList<Integer>();
    }

    public boolean TryGetValue(Character c,   TrieNodeEx node)
    {
        if (minflag <=  c && maxflag >=  c) {
            if( m_values.containsKey(c)){
                node=m_values.get(c);
                return true;
            }
            return false;           }
        node = null;
        return false;
    }

    public TrieNodeEx Add(Character c)
    {
        if( m_values.containsKey(c)){
            return m_values.get(c);
        }
        if (minflag > c) { minflag = (int)c; }
        if (maxflag < c) { maxflag = (int)c; }

        TrieNodeEx node = new TrieNodeEx();
        node.Parent = this;
        node.Char = c;
        m_values.put(c, node);
        Count++;
        return node;
    }

    public void SetResults(int text)
    {
        if (End == false) {
            End = true;
        }
        if (Results.contains(text) == false) {
            Results.add(text);
        }
    }

    public void Merge(TrieNodeEx node)
    {
        TrieNodeEx nd = node;
        while ((int)nd.Char !=  0) {
            node.m_values.forEach((key,value)->{
                if(m_values.containsKey(key)==false){
                    if(merge_values.containsKey(key)==false){
                        if (minflag > key) { minflag = (int)key; }
                        if (maxflag < key) { maxflag = (int)key; }
                        merge_values.put(key, value);
                        Count++;
                    }
                }
            });
            nd = nd.Failure;
        }
    }

    public int Rank(TrieNodeEx[] has)
    {
        boolean[] seats = new boolean[has.length];
        Integer  start = 1;

        has[0] = this;

        Rank(start, seats, has);
        int maxCount = has.length - 1;
        while (has[maxCount] == null) { maxCount--; }
        return maxCount;
    }

    private void Rank(Integer  start, boolean[] seats, TrieNodeEx[] has)
    {
        if (maxflag == 0) return;

        List<Integer> keys=new ArrayList<Integer>();
        m_values.forEach((k,v)->{ keys.add((int)k); });   
        merge_values.forEach((k,v)->{ keys.add((int)k); });

        while (has[start] != null) { start++; }

        int s = start < (int)minflag ? (int)minflag : start;

        for (int i = s; i < has.length; i++) {
            if (has[i] == null) {
                Integer next = i - (int)minflag;
                //if (next < 0) continue;
                if (seats[next]) continue;

                boolean isok = true;
                for (Integer item : keys) {
                    if (has[next + item] != null) { isok = false; break; }
                }
   
                if (isok) {
                    SetSeats(next, seats, has);
                    break;
                }
            }
        }
        start += keys.size() / 2;

        //var keys2 = m_values.OrderByDescending(q => q.Value.Count).ThenByDescending(q => q.Value.maxflag - q.Value.minflag);


        for (Character key : m_values.keySet()) {
            TrieNodeEx value=m_values.get(key);
            value.Rank(start, seats, has);
        }
    }


    private void SetSeats(Integer next, boolean[] seats, TrieNodeEx[] has)
    {
        Next = next;
        seats[next] = true;

        merge_values.forEach((key,value)->{
            int position = next + key;
            has[position] = value;
        });

         m_values.forEach((key,value)->{
            int position = next + key;
            has[position] = value;
        });
    }



}

File: golang/ToolGood/Words/internals/TrieNodeEx.go
package internals

type TrieNodeEx struct {
	Parent *TrieNodeEx  
	Failure *TrieNodeEx  
	Char int32  
	End bool   
	Results []int 
	M_values map[int32]*TrieNodeEx  
	Merge_values map[int32]*TrieNodeEx  
	minflag int32 
	maxflag int32
	Next int  
	Count int 
}

func NewTrieNodeEx() *TrieNodeEx  {
	return &TrieNodeEx{
		M_values: make(map[int32]*TrieNodeEx, 0),
		Merge_values: make(map[int32]*TrieNodeEx, 0),
		Results: make([]int, 0),
		minflag: 0xffff,
		maxflag:0,
		Next:0,
		Count:0,
	}
}

func (this *TrieNodeEx) TryGetValue(c int32) (bool,*TrieNodeEx){
	if this.minflag<=c && this.maxflag>=c {
		if val, s := this.M_values[c]; s {
			return true,val
		}
	}
	return false,nil
}


func (this *TrieNodeEx) Add(c int32) *TrieNodeEx{
	if val, s := this.M_values[c]; s {
		return val
	}
	if this.minflag>c {
		this.minflag = c
	}
	if this.maxflag<c {
		this.maxflag = c
	}
	node := NewTrieNodeEx()
	node.Parent = this;
	node.Char = c;
	this.M_values[c]=node
	this.Count++;
	return node 
}

func (this *TrieNodeEx) SetResults(text int) {
	if this.End==false {
		this.End=true;
	}
	for i := 0; i < len(this.Results); i++ {
        if this.Results[i] == text {
            return
        }
    }
	this.Results=append(this.Results,text)
}

func (this *TrieNodeEx)Merge(node *TrieNodeEx) {
	nd:=node
	for	nd.Char != 0 {
		for key,value := range node.M_values{
			if _,s := this.M_values[key]; s{
				continue;
			} 
			if _,s := this.Merge_values[key]; s{
				continue;
			} 
			if this.minflag>key {
				this.minflag = key
			}
			if this.maxflag<key {
				this.maxflag = key
			}
			this.Merge_values[key]=value
			this.Count++;
		}
		nd = nd.Failure;
	}
}

func (this *TrieNodeEx)Rank(has []*TrieNodeEx) int {
    var seats []bool =make([]bool,len(has))
	var start int = 1
	has[0] = this;
	this.Rank2(start, seats, has)
	maxCount := len(has) - 1;
	for	has[maxCount] == nil {
		maxCount--;
	}
	return maxCount;
}
 
func (this *TrieNodeEx)Rank2(start int,seats []bool,has []*TrieNodeEx) int{
	if (this.maxflag == 0) {
		return start
	}
	keys := make([]int32,0)
	for k,_:=range this.M_values {
		keys=append(keys,k)
	}
	for k,_:=range this.Merge_values {
		keys=append(keys,k)
	}	

	for	has[start] != nil{
		start++
	}
	s := start
	if start < int (this.minflag){
		s= int (this.minflag)
	}

	for i := s; i < len(has); i++{
		if has[i]==nil {
			next := i - int (this.minflag);
			if seats[next]{
				continue
			}
			isok := true;

			for	_,item :=range keys{
				if has[next+ int (item)] != nil{
					isok = false
					break
				}
			}
			if (isok) {
				this.SetSeats(next, seats, has);
				break;
			}
		}
	}
	start += len(keys) / 2;
	
	for _,value:= range this.M_values{
		start=value.Rank2(start, seats, has);
	}
	return start
}

func (this *TrieNodeEx)SetSeats(next int,seats []bool,has []*TrieNodeEx){
	this.Next = next;
	seats[next] = true;
	for key,value := range this.Merge_values{
		position := next + int (key);
		has[position] = value;
	}
	for key,value := range this.M_values{
		position := next + int (key) ;
		has[position] = value;
	}
}

 